primitives/trie/src/storage_proof.rs

```diff
@@ -14,65 +14,990 @@
+/// These do not need to be part of the proof but are required
+/// when processing the proof.
+pub enum Input {
+	/// Proof is self contained.
+	None,
+
+	/// Contains trie roots used during proof processing.
+	ChildTrieRoots(ChildrenProofMap<Vec<u8>>),
+
+	/// Contains trie roots used during proof processing.
+	/// Contains key and values queried during the proof processing.
+	QueryPlanWithValues(ChildrenProofMap<(Vec<u8>, Vec<(Vec<u8>, Option<Vec<u8>>)>)>),
+
+	/// Contains trie roots used during proof processing.
+	/// Contains keys queried during the proof processing.
+	QueryPlan(ChildrenProofMap<(Vec<u8>, Vec<Vec<u8>>)>),
+}
+
+impl Input {
+	/// Update input with new content.
+	/// Return false on failure.
+	/// Fails when the input type differs, except for `None` input
+	/// that is always reassignable.
+	///
+	/// Not that currently query plan inputs are not mergeable
+	/// even if doable (just unimplemented).
+	#[must_use]
+	pub fn consolidate(&mut self, other: Self) -> bool {
+		match self {
+			Input::None => {
+				*self = other;
+			},
+			Input::ChildTrieRoots(children) => {
+				match other {
+					Input::None => (),
+					Input::ChildTrieRoots(children_other) => {
+						for (child_info, root) in children_other {
+							match children.entry(child_info) {
+								btree_map::Entry::Occupied(v) => if v.get() != &root {
+									return false;
+								},
+								btree_map::Entry::Vacant(v) => {
+									v.insert(root);
+								},
+							}
+						}
+					},
+					Input::QueryPlan(..) => return false,
+					Input::QueryPlanWithValues(..) => return false,
+				}
+			},
+			Input::QueryPlan(..) => return false,
+			Input::QueryPlanWithValues(..) => return false,
+		}
+		true
+	}
+}
+
+/// Kind for a `Input` variant.
+pub enum InputKind {
+	/// `Input::None` kind.
+	None,
+
+	/// `Input::ChildTrieRoots` kind.
+	ChildTrieRoots,
+
+	/// `Input::QueryPlan` kind.
+	QueryPlan,
+
+	/// `Input::QueryPlanWithValues` kind.
+	QueryPlanWithValues,
+}
+
+impl StorageProofKind {
+	/// Input kind needed for processing (create) the proof.
+	pub fn process_input_kind(&self) -> InputKind {
+		match self {
+			StorageProofKind::KnownQueryPlanAndValues => InputKind::QueryPlan,
+			StorageProofKind::TrieSkipHashesForMerge
+				| StorageProofKind::TrieSkipHashes
+				| StorageProofKind::TrieSkipHashesFull => InputKind::ChildTrieRoots,
+			StorageProofKind::Full
+				| StorageProofKind::Flatten => InputKind::None,
+		}
+	}
+
+	/// Input kind needed for verifying the proof.
+	pub fn verify_input_kind(&self) -> InputKind {
+		match self {
+			StorageProofKind::KnownQueryPlanAndValues => InputKind::QueryPlanWithValues,
+			StorageProofKind::TrieSkipHashes
+				| StorageProofKind::TrieSkipHashesFull
+				| StorageProofKind::TrieSkipHashesForMerge
+				| StorageProofKind::Full
+				| StorageProofKind::Flatten => InputKind::None,
+		}
+	}
+
+	/// Indicates what variant of proof recorder should be use.
+	pub fn need_register_full(&self) -> bool {
```

Comment:
This method name sounds like it's 3 random words, I think with a bit of care it can be rephrased to something a bit more self-explanatory, like;
`is_full_proof_recorder_needed()`